### PR TITLE
feat: add "unlisted" front matter option for blog posts

### LINKED ISSUE #1393: Add option to unlist blog post from sidebar
## 🚀 Feature

It would be nice if it was possible to unlist a blog post so that it doesn't appear in the sidebar.

For example, a blog post with the frontmatter below would not appear in the sidebar.

```md
title: My Unlisted Blog Post
unlisted: true
---
Hello world!
```

I tried looking for such a header option at https://docusaurus.io/docs/en/adding-blog#header-options but I couldn't any similar functionality.

## Have you read the [Contributing Guidelines on issues](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#reporting-new-issues)?

Yes.

## Motivation

The motivation for this feature is to make it possible to change dates of published posts without breaking the old URL (which might be widely shared).

Concrete example, we published a blog post two weeks ago that is dated for 2020 (https://scalameta.org/metals/blog/2020/04/12/mercury.html). We have closed three PRs fixing the date (https://github.com/scalameta/metals/pull/683, https://github.com/scalameta/metals/pull/670, https://github.com/scalameta/metals/pull/656) but they have all been closed because changing the date of the post would turn the old URL into a 404.

With the ability to unlist posts, I could unlist the 2020 post, create a new posts dated in 2019 and update the unlisted 2020 post to forward to the new 2019 post.

I'm open for alternative solutions to fix that particular problem. I think unlisting is a flexible solution that could be useful in other situations as well.

## Pitch

Docusaurus blogs are great! Adding the ability to unlist a blog post would make blogs even better by providing flexibility in how to organize the blog sidebar.

### PR DESCRIPTION
**Pull request by olafurpg**
Previously, the blog sidebar listed the most recent blog posts sorted by
their publish date. This commit adds the ability to hide a specific blog
post from the sidebar and the feed of all blog posts by adding the
`unlisted: true` header option.

Fixes #1393.

## Motivation

The motivation for this feature is to make it possible to change dates of published posts without breaking the old URL (which might be widely shared).

Concrete example, we published a blog post two weeks ago that is dated for 2020 (https://scalameta.org/metals/blog/2020/04/12/mercury.html). We have closed three PRs fixing the date (https://github.com/scalameta/metals/pull/683, https://github.com/scalameta/metals/pull/670, https://github.com/scalameta/metals/pull/656) but they have all been closed because changing the date of the post would turn the old URL into a 404.

With the ability to unlist posts, I could unlist the 2020 post, create a new posts dated in 2019 and update the unlisted 2020 post to forward to the new 2019 post.

I'm open for alternative solutions to fix that particular problem. I think unlisting is a flexible solution that could be useful in other situations as well, for example when writing a draft post.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebook/Docusaurus/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

I would appreciate help for testing this change since I'm unfamiliar with jest (and JavaScript in general 😅 ). I created a blog post in `__test__/__fixtures__` using `unlisted: true` and ran `yarn test` and everything passed but I expected some snapshot tests to fail.

I tested the change manually on our website and it seems to work as expected:

Observe that uncommenting `unlisted: true` makes the post disappear from both the sidebar and the list of all blog posts while the post is still published under the unchanged URL.

## Related PRs

I'm not aware of any.

### CODE REVIEW

File: packages/docusaurus-1.x/lib/core/BlogPageLayout.js [hunk 1]
```diff
@@ -10,6 +10,8 @@ const BlogPost = require('./BlogPost.js');
 const BlogSidebar = require('./BlogSidebar.js');
 const Container = require('./Container.js');
 const MetadataBlog = require('./MetadataBlog.js');
+
+const MetadataPublicBlog = MetadataBlog.filter(item => item.unlisted !== true);
```
Comment (olafurpg) [inline packages/docusaurus-1.x/lib/core/BlogPageLayout.js]: Where would be a good place to put this shared logic between `BlogSidebar.js`, `BlogPageLayout.js` and `feed.js`?

File: packages/docusaurus-1.x/lib/core/BlogPageLayout.js [hunk 2]
```diff
@@ -40,27 +42,28 @@ class BlogPageLayout extends React.Component {
           />
           <Container className="mainContainer postContainer blogContainer">
             <div className="posts">
-              {MetadataBlog.slice(page * perPage, (page + 1) * perPage).map(
```
Comment (olafurpg) [inline packages/docusaurus-1.x/lib/core/BlogPageLayout.js]: This diff is easier on the eye with "No whitespace" enabled.

File: packages/docusaurus-1.x/lib/core/BlogPageLayout.js (hunk 1 above)
Comment (gabro) [inline packages/docusaurus-1.x/lib/core/BlogPageLayout.js]:
```suggestion
const MetadataPublicBlog = MetadataBlog.filter(item => !item.unlisted);
```

File: packages/docusaurus-1.x/lib/server/feed.js [hunk 3]
```diff
@@ -28,6 +28,9 @@ module.exports = function(type) {
 
   readMetadata.generateMetadataBlog();
   const MetadataBlog = require('../core/MetadataBlog.js');
+  const MetadataPublicBlog = MetadataBlog.filter(
+    item => item.unlisted !== true,
```
Comment (NishealJ) [inline packages/docusaurus-1.x/lib/server/feed.js]: item => !item.unlisted,

**Comment by olafurpg** [comment]
Thank you for the review! I've renamed "unlisted" into "draft" and incorporated the `!item.draft` refactoring. Note that there is a difference between draft posts in Jekyll and how they are implemented in this PR. 

- Jekyll draft posts are ignored during a non-development build and they do not need a date in the filename
- draft posts in this PR still require a date in the filename and are published during non-development builds, they just don't appear in the blog sidebar or main blog feed

> Is there currently a way for you to view your unlisted/draft posts without toggling the flag?

Not in the current implementation. It would be nice to see draft posts during livereload development but I'm not sure how to implement that.

File: packages/docusaurus-1.x/lib/core/BlogPageLayout.js (hunk 1 above)
Comment (olafurpg) [reply packages/docusaurus-1.x/lib/core/BlogPageLayout.js]: Done.

File: packages/docusaurus-1.x/lib/server/feed.js (hunk 3 above)
Comment (olafurpg) [reply packages/docusaurus-1.x/lib/server/feed.js]: Done.

**Comment by gabro** [comment]
About unlisted vs draft. The way I would expect "draft" to behave, is not to make the post public, whereas unlisted would be published but not listed in the menu.

To summarize

 status  | visible in menu | published |
---------|:---------------:|:---------:|
draft    |      no         |   no      |
unlisted |      no         |   yes     |
normal   |      yes        |   yes     |

For this reason, I think 'unlisted' was the appropriate name for the changes included in this PR.

**Comment by olafurpg** [comment]
Anything I can do to help unblock this issue? All of the review comments from https://github.com/facebook/Docusaurus/pull/1396#pullrequestreview-231255400 have been addressed, although I believe it's worth considering using `unlisted` instead of `draft` due to the reasons mentioned by @gabro in https://github.com/facebook/Docusaurus/pull/1396#issuecomment-487561180

**Comment by endiliey** [comment]
cc @yangshun @wgao19

**Pull request merged**